parlai/crowdsourcing/tasks/model_chat/analysis/compile_results.py

```diff
@@ -0,0 +1,522 @@
+        self.results_folders = opt['results_folders'].split(',')
+        assert len(self.results_folders) > 0
+        for folder in self.results_folders:
+            assert os.path.isdir(folder), f'{folder} is not a valid folder!'
+        self.output_folder = opt['output_folder']
+        os.makedirs(self.output_folder, exist_ok=True)
+        self.start_date = opt['start_date']
+        self.max_convos_per_worker = opt['max_convos_per_worker']
+        self.min_word_count = opt['min_word_count']
+        self.hit_block_list = opt['hit_block_list'].split(',')
+        self.worker_block_list = opt['worker_block_list'].split(',')
+
+        # Setting up problem buckets
+        if 'none_all_good' not in self.PROBLEM_BUCKETS:
+            raise ValueError(
+                'There must be a "none_all_good" category in self.PROBLEM_BUCKETS!'
+            )
+        self.regular_buckets = [
+            bucket
+            for bucket in self.PROBLEM_BUCKETS
+            if bucket not in ['other', 'none_all_good']
+        ]
+        # Remove the buckets that are special cases
+
+        self.acceptability_checker = AcceptabilityChecker()
+
+    def compile_results(self) -> pd.DataFrame:
+
+        read_folders = []
+        date_strings = []
+        for folder in self.results_folders:
+            # Load paths
+            date_strings = sorted(
+                [
+                    obj
+                    for obj in os.listdir(folder)
+                    if os.path.isdir(os.path.join(folder, obj))
+                    and re.fullmatch(r'\d\d\d\d_\d\d_\d\d', obj)
+                ]
+            )
+            if self.start_date != '':
+                date_strings = [
+                    str_ for str_ in date_strings if str_ >= self.start_date
+                ]
+            folders = [os.path.join(folder, str_) for str_ in date_strings]
+            read_folders.extend(folders)
+        print(f'Date folders: ' + ', '.join(date_strings))
+
+        now = datetime.now()
+        results_file = os.path.join(
+            self.output_folder,
+            f'results_{"".join([d + "_" for d in date_strings])[:-1]}_{now.strftime("%Y%m%d_%H%M%S")}.csv',
+        )
+        worker_results_file = os.path.join(
+            self.output_folder,
+            f'worker_results_{"".join([d + "_" for d in date_strings])[:-1]}_{now.strftime("%Y%m%d_%H%M%S")}.csv',
+        )
+        # Read in each file
+        num_incomplete_convos = 0
+        num_complete_convos = 0
+        complete_convos_per_model = {}
+        bad_conversations = []
+        old_experimental_design_conversations = []
+        problem_counts = {}
+        worker_stats = {}
+        worker_conversation_counts = {}
+        total_utterances = 0
+
+        conversation_idx = 0
+        conversation_dfs = []
+        for read_folder in read_folders:
+            read_folder_name = os.path.split(read_folder)[-1]
+            for file_name in sorted(os.listdir(read_folder)):
```

Comment:
why do we sort here?

Reply:
My guess is by sorting the file name( which is equivalent to sorting by datetime the HIT is submitted), it can process data in chronological order? For example L183 seems to be skipping any later submission if the turker has done certain amount.

Reply:
Ah, the reason is much more boring - a stable sort order guarantees reproducible results for unit tests